### PR TITLE
feat: add qbraid_qir.qiskit module for Qiskit to QIR conversion

### LINKED ISSUE #272: Add qbraid_qir.qiskit module (port of archived qiskit-qir)
## Summary

The [microsoft/qiskit-qir](https://github.com/microsoft/qiskit-qir) package has been archived and is incompatible with Qiskit 2.x. The qBraid SDK currently depends on this package for `qiskit → pyqir` conversions.

## Proposal

Port the qiskit-qir functionality into qbraid-qir as a new `qbraid_qir.qiskit` module:

- `qiskit_to_qir(circuit, name, **kwargs)` — main conversion function
- Updated for Qiskit 2.x compatibility
- Follows existing qbraid-qir conventions (`cirq_to_qir`, `qasm3_to_qir` pattern)
- Supports pyqir 0.10-0.12+ (typed and opaque pointers)

## Motivation

1. **Archived dependency**: qiskit-qir is no longer maintained
2. **Qiskit 2.x**: The old package requires `qiskit<2.0`
3. **Consolidation**: Keep all QIR conversion code under qbraid-qir

## Related

- **Implementation PR**: #271
- **SDK integration PR**: [qBraid/qBraid#1132](https://github.com/qBraid/qBraid/pull/1132)

## License

The original qiskit-qir is MIT licensed. Attribution will be included per Apache 2.0 + MIT compatibility requirements.

### PR DESCRIPTION
**Pull request by eidrynbot**
## Summary

Port of [microsoft/qiskit-qir](https://github.com/microsoft/qiskit-qir) (MIT License, now archived) into qbraid-qir as `qbraid_qir.qiskit`.

Closes #272

### New Module: `qbraid_qir.qiskit`
- **`qiskit_to_qir(circuit, name, **kwargs)`**: Main conversion function matching the `cirq_to_qir`/`qasm3_to_qir` interface
- **`QiskitModule`**: Circuit wrapper for QIR conversion
- **`BasicQiskitVisitor`**: Visitor for converting Qiskit operations to QIR
- **`QiskitConversionError`**: Exception for conversion failures

### Key Updates from Original qiskit-qir
- **Qiskit 2.x compatibility**: Updated for new `QuantumCircuit` API (`CircuitInstruction` instead of tuples)
- **pyqir 0.10-0.12+ support**: Works with both typed pointers (0.10.x) and opaque pointers (0.12+)
- **qbraid-qir conventions**: Follows existing code style and patterns

### Testing
- 44 new unit tests covering gates, measurements, composite gates, error cases
- All tests pass with existing test suite (225 passed, 5 skipped)

### Documentation
- Updated README with qiskit installation and usage example
- Added API reference page for `qbraid_qir.qiskit`
- Added `NOTICE.md` for MIT license attribution

### Configuration
- Added `qiskit` optional dependency in pyproject.toml
- Updated tox.ini test environments

## License Compliance
The original qiskit-qir is MIT licensed. Code derived from it includes proper attribution in file headers (following the pattern from qBraid SDK) and full license text in `NOTICE.md`.

## Related
- **Tracking issue**: #272
- **SDK integration PR**: [qBraid/qBraid#1132](https://github.com/qBraid/qBraid/pull/1132)

## Summary by CodeRabbit

* **New Features**
  * Added Qiskit → QIR conversion with a public conversion entrypoint, optional transpilation, and runtime/output/barrier controls.

* **Documentation**
  * Added README installation/example, API docs page, and NOTICE with third‑party attribution and license.

* **Tests**
  * Added comprehensive unit and integration tests covering many Qiskit circuits, edge cases, transpilation, and error handling.

* **Chores**
  * Added Qiskit as an optional extra, updated packaging metadata and CI/test install to include it.

[Review Change Stack](https://app.coderabbit.ai/change-stack/qBraid/qbraid-qir/pull/271?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by ryanhill1** [comment]
@TheGupta2012 FYI:

@eidrynbot Is one of my bots that I have working on a few things

File: qbraid_qir/qiskit/NOTICE.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Third Party Notices
+
+This software includes code derived from third party sources.
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/NOTICE.md:3]: "Third Party Notices" is a title/noun phrase where hyphenation is optional (AP style). There is no markdownlint in this project's CI pipeline, so MD032/MD044 rules are not enforced. No change needed.

File: qbraid_qir/qiskit/NOTICE.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE
+```
+
+The qiskit-qir codebase has been adapted to:
+- Work with Qiskit 2.x (updated for new QuantumCircuit API)
+- Follow qbraid-qir conventions and coding style
+- Support both pyqir 0.10.x and 0.12+ (typed and opaque pointers)
+- Integrate with the qbraid-qir module structure
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/NOTICE.md:37]: The file already ends with exactly one trailing newline (verified with `od -c`). No change needed.

File: tests/qiskit_qir/conftest.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+"""
+Pytest fixtures for Qiskit QIR tests.
+
+"""
+
+import pytest
+from qiskit import ClassicalRegister, QuantumCircuit, QuantumRegister
+
+
+@pytest.fixture()
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply tests/qiskit_qir/conftest.py:24]: Ruff PT001 is not part of this project's CI pipeline (which uses pylint + isort + black + mypy). All 13 fixtures across the test suite consistently use `@pytest.fixture()` with parentheses — this is the established convention. No change needed.

File: tests/qiskit_qir/test_basic_gates.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    "cx": "cnot",
+    "cz": "cz",
+    "swap": "swap",
+}
+
+
+class TestSingleQubitGates:
+    """Tests for single qubit gate conversions."""
+
+    @pytest.mark.parametrize("gate_name,qir_name", SINGLE_QUBIT_GATES.items())
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply tests/qiskit_qir/test_basic_gates.py]: Ruff PT006 is not enforced in this project's CI pipeline. The comma-separated string format `"gate_name,qir_name"` is standard pytest and the predominant pattern across the codebase. No change needed.

File: tests/qiskit_qir/test_qiskit_to_qir.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+# ---------------------------------------------------------------------------
+# Error handling
+# ---------------------------------------------------------------------------
+
+
+class TestQiskitToQirErrors:
+    """Tests for error handling in qiskit_to_qir."""
+
+    def test_invalid_input_type(self):
+        with pytest.raises(TypeError, match="must be of type qiskit.QuantumCircuit"):
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply tests/qiskit_qir/test_qiskit_to_qir.py:216]: Ruff RUF043 is not enforced in this project's CI pipeline. The `.` in `qiskit.QuantumCircuit` has zero realistic false-positive risk in this `pytest.raises(match=...)` context. No change needed.

File: qbraid_qir/qiskit/__init__.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+   QiskitConversionError
+
+"""
+
+from .convert import qiskit_to_qir
+from .elements import QiskitModule
+from .exceptions import QiskitConversionError
+from .visitor import BasicQiskitVisitor
+
+__all__ = ["qiskit_to_qir", "QiskitModule", "QiskitConversionError", "BasicQiskitVisitor"]
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/__init__.py:59]: Ruff RUF022 is not enforced in this project's CI pipeline (which uses pylint + isort + black + mypy). The current `__all__` order follows logical grouping matching the module docstring. No change needed.

File: qbraid_qir/qiskit/convert.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+from pyqir import Context, Module, qir_module
+from qiskit.circuit import QuantumCircuit
+
+from .elements import QiskitModule, generate_module_id
+from .exceptions import QiskitConversionError
+from .visitor import BasicQiskitVisitor
+
+
+def qiskit_to_qir(circuit: QuantumCircuit, name: Optional[str] = None, **kwargs) -> Module:
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/convert.py]: The kwargs are documented in the docstring with types. Adding a `TypedDict` for 3 boolean kwargs (`record_output`, `emit_barrier_calls`, `transpile`) is over-engineering for an internal forwarding function. No change needed.

File: qbraid_qir/qiskit/visitor.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        self._module: pyqir.Module
+        self._qiskit_module: QiskitModule | None = None
+        self._builder: pyqir.Builder
+        self._entry_point: str = ""
+        self._qubit_labels: dict[Qubit, int] = {}
+        self._clbit_labels: dict[Clbit, int] = {}
+        self._measured_qubits: dict[int, bool] = {}
+        self._initialize_runtime = initialize_runtime
+        self._record_output = record_output
+        self._emit_barrier_calls = emit_barrier_calls
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/visitor.py]: Already suppressed with `# pylint: disable=too-many-instance-attributes` on line 74. CI passes clean (pylint rates 10/10). This comment is stale — the issue was addressed before the fix commit.

File: qbraid_qir/qiskit/elements.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+    @property
+    def num_clbits(self) -> int:
+        """Return the number of classical bits."""
+        return self._num_clbits
+
+    @classmethod
+    def from_circuit(
+        cls, circuit: QuantumCircuit, module: Optional[Module] = None
+    ) -> "QiskitModule":
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/elements.py:149]: The `None` default exists for flexibility. The public API (`qiskit_to_qir`) always constructs and passes a module. The visitor now raises `ValueError` with a clear message if `module.module is None` (fixed in d307221), providing a proper error path. No further defensive code needed in `from_circuit`.

File: docs/api/qbraid_qir.qiskit.rst [hunk 10]
```diff
@@ -0,0 +1,8 @@
+:orphan:
+
+qbraid_qir.qiskit
+==================
+
+.. automodule:: qbraid_qir.qiskit
+   :undoc-members:
+   :show-inheritance:
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply docs/api/qbraid_qir.qiskit.rst:8]: All 3 existing rst files (cirq, qasm3, squin) use the same `:undoc-members: :show-inheritance:` pattern without `:members:`. This is a project-wide convention. Changing it for just the qiskit module would be inconsistent; changing all 4 is out of scope for this PR.

File: qbraid_qir/__init__.py [hunk 11]
```diff
@@ -51,13 +51,15 @@
     "dumps",
     "qasm3_to_qir",
     "cirq_to_qir",
+    "qiskit_to_qir",
 ]
 
-_lazy = {"cirq": "cirq_to_qir", "qasm3": "qasm3_to_qir"}
+_lazy = {"cirq": ("cirq_to_qir",), "qasm3": ("qasm3_to_qir",), "qiskit": ("qiskit_to_qir",)}
```
Comment (ryanhill1) [inline qbraid_qir/__init__.py:57]: why convert to tuples? I believe this might break the lazy importing logic

File: qbraid_qir/qiskit/visitor.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+class QuantumCircuitElementVisitor(metaclass=ABCMeta):
+    """Abstract base class for quantum circuit element visitors."""
+
+    @abstractmethod
+    def visit_register(self, register):
+        """Visit a register element."""
+
+    @abstractmethod
+    def visit_instruction(self, instruction, qargs, cargs):
```
[the remark this answers is not in the document]
Comment (ryanhill1) [reply qbraid_qir/qiskit/visitor.py]: @TheGupta2012 I agree, can you add type hints here?

File: qbraid_qir/qiskit/visitor.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+        for size in module.reg_sizes:
+            rt.array_record_output(
+                self._builder,
+                pyqir.const(IntType(self._module.context, 64), size),
+                Constant.null(i8p),
+            )
+            for index in range(size - 1, -1, -1):
+                result_ref = pyqir.result(self._module.context, logical_id_base + index)
+                rt.result_record_output(self._builder, result_ref, Constant.null(i8p))
+            logical_id_base += size
```
[the remark this answers is not in the document]
Comment (ryanhill1) [reply qbraid_qir/qiskit/visitor.py]: @TheGupta2012 Can you investigate and determine if this is actually an issue?

File: qbraid_qir/qiskit/NOTICE.md
Comment (ryanhill1) [inline qbraid_qir/qiskit/NOTICE.md:1]: Let's move this to `qbraid_qir/qiskit/NOTICE.md` instead

**Review by ryanhill1 — CHANGES_REQUESTED** [review]

File: qbraid_qir/qiskit/visitor.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qiskit/visitor.py]: Yeah indeed it is and the design is stemming from the `qbraid_qir/profiles/core.py` where each profile had a specific `visitor` and `module` in the function signature - 

https://github.com/qBraid/qbraid-qir/blob/8de71bb1defaff4c33d74670588b991f73d59941/qbraid_qir/profiles/core.py#L128


However, for the qiskit visitor this might be overkill since we do not have any profiles per se - will revert it to just refer to `self._module` inside the method

File: qbraid_qir/__init__.py (hunk 11 above)
Comment (TheGupta2012) [reply qbraid_qir/__init__.py:57]: The tuples are intentional and do not break lazy imports. The original code used plain strings:

```python
_lazy = {"cirq": "cirq_to_qir", "qasm3": "qasm3_to_qir", "qiskit": "qiskit_to_qir"}
```

In `__getattr__`, the check `name in objects` performs **substring** matching on strings (e.g. `"qir" in "cirq_to_qir"` is `True`), which works by accident here but is a latent bug. With tuples, `name in ("cirq_to_qir",)` performs proper **element** membership — the correct semantics. All tests and lazy import behavior are unchanged.

Comment (TheGupta2012) [reply qbraid_qir/__init__.py:57]: Also, the following code produces a cryptic error in current implementation - 

```python 
In [5]: import qbraid_qir

In [6]: qbraid_qir.qir
-------------------------------------------------------------
AttributeError              Traceback (most recent call last)
Cell In[6], line 1
----> 1 qbraid_qir.qir

File ~/Desktop/qBraid/repos/qiskit-qir-updates/qbraid-qir/qbraid_qir/__init__.py:74, in __getattr__(name)
     72 if name in objects:
     73     module = importlib.import_module(f".{mod_name}", __name__)
---> 74     obj = getattr(module, name)
     75     globals()[name] = obj
     76     return obj

AttributeError: module 'qbraid_qir.cirq' has no attribute 'qir'
``` 

when the error should be - 

```python
AttributeError: module 'qbraid_qir' has no attribute 'qir'
```

**Review by ryanhill1 — APPROVED** [review]

**Review by TheGupta2012 — APPROVED** [review]
